src/mame/konami/konami_gn676_lan.cpp

```diff
@@ -65,176 +89,287 @@ This pcb is the same as the A version but with one added chip:
+	: konami_gn676_lan_device(mconfig, KONAMI_GN676B_LAN, tag, owner, clock)
+{
+}
+
+void konami_gn676b_lan_device::device_add_mconfig(machine_config &config)
+{
+	X76F041(config, m_x76f041);
+}
 
 void konami_gn676_lan_device::device_start()
 {
 	m_lanc2_ram = std::make_unique<uint8_t[]>(0x8000);
 
-	save_item(NAME(m_fpga_uploaded));
 	save_item(NAME(m_lanc2_ram_r));
 	save_item(NAME(m_lanc2_ram_w));
 	save_item(NAME(m_lanc2_reg));
+	save_item(NAME(m_network_buffer_max_size));
+	save_item(NAME(m_network_id));
+	save_item(NAME(m_fpga_reset_state));
+	save_item(NAME(m_fpga_uploaded));
+	save_item(NAME(m_fpga_waiting_firmware));
+	save_item(NAME(m_fpga_receiving));
+	save_item(NAME(m_fpga_is_stubbed));
+	save_item(NAME(m_fpga_firmware_size));
+	save_item(NAME(m_fpga_firmware_crc));
+	save_item(NAME(m_x76f041_enabled));
+	save_item(NAME(m_x76f041_read_enabled));
+	save_item(NAME(m_x76f041_rst_triggered));
+
 	save_pointer(NAME(m_lanc2_ram), 0x8000);
 }
+void konami_gn676_lan_device::device_reset()
+{
+	std::fill_n(m_lanc2_ram.get(), 0x8000, 0);
 
+	m_fpga_reset_state = true;
+	reset_fpga_state(false);
+}
 
-uint32_t konami_gn676_lan_device::lanc1_r(offs_t offset)
+uint8_t konami_gn676_lan_device::lanc1_r(offs_t offset)
 {
-	switch (offset)
+	uint8_t r = 0xff;
+
+	if (offset == 0x40)
 	{
-		case 0x40/4:
-		{
-			uint32_t r = 0;
+		r = (m_fpga_receiving << 6) | (m_fpga_waiting_firmware << 5);
 
-			r |= (m_fpga_uploaded) ? (1 << 6) : 0;
-			r |= 1 << 5;
+		// racingj/racingj2 polls expecting to see bit 5 set and bit 6 unset before
+		// it'll start uploading the firmware, so only set it after being polled.
+		m_fpga_receiving = true;
+	}
 
-			return (r) << 24;
-		}
+	LOG("%s: lanc1_r: read %02X from %08X\n", tag(), r, offset, machine().describe_context());
 
-		default:
+	return r;
+}
+
+void konami_gn676_lan_device::lanc1_w(offs_t offset, uint8_t data)
+{
+	LOG("%s: lanc1_w: wrote %02X to %08X\n", machine().describe_context(), data, offset);
+
+	if (offset == 0x20)
+	{
+		if (BIT(data, 7))
 		{
-			LOG("lanc1_r: %08X at %08X\n", offset, machine().describe_context());
-			return 0xffffffff;
+			m_fpga_uploaded = true;
+			m_fpga_waiting_firmware = false;
+			m_fpga_receiving = false;
+			m_fpga_firmware_crc = util::crc32_creator::simple(m_lanc2_ram.get(), m_fpga_firmware_size);
+
+			LOG("%s: lanc1_fpga: Found firmware with hash of %08x\n", tag(), m_fpga_firmware_crc, machine().describe_context());
+
+			m_x76f041_enabled = m_fpga_firmware_crc == 0xa8c97a75 || m_fpga_firmware_crc == 0xdfc74cc9 || m_fpga_firmware_crc == 0x93b86e35;
+			m_fpga_is_stubbed = m_fpga_firmware_crc == 0x3760e3ce;
+
+			/*
+			// DEBUG: Dump firmware for quick comparison
+			FILE *f = fopen(util::string_format("firmware_%08x.bin", m_fpga_firmware_crc).c_str(), "wb");
+			if (f)
+			{
+				fwrite(m_lanc2_ram.get(), 1, m_fpga_firmware_size, f);
+				fclose(f);
+			}
+			*/
```

Comment:
Can you use `if (SOME_MACRO_DEFINED_TO_ZERO_BY_DEFAULT)` to disable this so the code can’t rot as easily?

Reply:
Sure. I'll keep that in mind if I need to add any other similar debug features in the future.